util/send-payload-pls.sh

```diff
@@ -1,17 +1,91 @@
 #!/bin/bash
 #
-# Script to post a payload against a local webserver at each paranoia level
+# Script to post a payload against a local webserver at each paranoia level.
 #
-# Note: Webserver has to be prepared to take desired PL as Request Header "PL"
+# Note: Webserver has to be prepared to take desired PL as Request Header "PL".
+
+# WARNING: Setting the paranoia level using a header without proper authentication and authorization
+# is extremely dangerous, and is not recommended for production.
+
 # Check the access log format at https://www.netnea.com/cms/apache-tutorial-5_extending-access-log/
+# This script assumes %{ModSecAnomalyScoreIn}e is the last access log column
+# For better results use the collaborative mode and set the SecDefaultAction to 'pass'
+# 
+# The anomaly score envvar can be set as follows:
+# SecAction "id:90101,phase:5,pass,nolog,setenv:ModSecAnomalyScoreIn=%{TX.anomaly_score}"
 #
+# Sample rule to setup the PL dinamically from localhost"
+#SecRule REMOTE_ADDR "@ipMatch 127.0.0.1,192.168.0.128" "id:90102,phase:1,pass,capture,log,auditlog,msg:'Setting engine to PL%{matched_var}',chain"
+#    SecRule REQUEST_HEADERS:PL "@rx ([1-4])" "setvar:'tx.executing_paranoia_level=%{matched_var}'"
+
 # Path to CRS rule set and local files
-CRS="/usr/share/modsecurity-crs/rules/"
+CRS="/usr/share/modsecurity-crs/rules"
 accesslog="/apache/logs/access.log"
 errorlog="/apache/logs/error.log"
+URL="localhost:40080"
+protocol="http"
+while [[ $# > 0 ]]
+do
+    case "$1" in
+        -c|--crs)
+            CRS="$2"
+            shift
+            ;;
+        -a|--access)
+            accesslog="$2"
+            shift
+            ;;
+        -e|--error)
+            errorlog="$2"
+            shift
+            ;;
+        -u|--url)
+            URL="$2"
+            shift
+            ;;
+        -r|--resolve)
+            resolve="$2"
+            resolve="--resolve $resolve"
+            shift
+            ;;
+        --protocol)
+            protocol="$2"
+            shift
+            ;;
+        --payload)
+            PAYLOAD="$2"
+            shift
+            ;;
+        -h|--help)
+            echo "Usage:"
+            echo " --access \"/apache/logs/access.log\""
+            echo " --error \"/apache/logs/error.log\""
+            echo " --url \"localhost:40080\""
+            echo " --resolve \"someservername:40080:localhost\""
+            echo " --protocol \"https\""
+            echo " --payload \"/tmp/payload\""
+            echo " --help"
+            exit 1
+            ;;
+    esac
+    shift
+done
+	    echo "Using CRS: $CRS"
+	    echo "Using accesslog: $accesslog"
+	    echo "Using errorlog: $errorlog"
+	    echo "Using URL: $URL"
+	    echo "Using protocol: $protocol"
+
+if [ -z "${PAYLOAD+x}" ]; then
```

Comment:
That does the `+x` mean here. I do not get this.

Reply:
This is just to avoid errors when checking the payload in case it is null